current/advanced/block-import.md

```diff
@@ -0,0 +1,41 @@
+
+The import queue is codified abstractly in Substrate by means of the [`ImportQueue` trait](https://substrate.dev/rustdocs/v2.0.0-alpha.8/sp_consensus/import_queue/trait.ImportQueue.html). The use of a trait allows each consensus engine to provide its own specialized implementation of the Import Queue which may take advantage of optimization opportunities such as verifying multiple blocks in parallel as they come in across the network.
+
+## The Basic Queue
+Substrate also provides a default implementation known as the [`BasicQueue`](https://substrate.dev/rustdocs/v2.0.0-alpha.8/sp_consensus/import_queue/struct.BasicQueue.html). The `BasicQueue` does not do any kind of optimization, rather it performs the verification and import steps sequentially. It does, however, abstract the notion of verification through the use of the `Verifier`(https://substrate.dev/rustdocs/v2.0.0-alpha.8/sp_consensus/import_queue/trait.Verifier.html) trait.
+
+Any consensus engine that relies on the `BasicQueue` must implement the `Verifier` trait. In the simplest cases, such as Sunstrate's instant seal consensus, the `Verifier` is simple declares all blocks as valid. In slot-based engines check that the blocks timestamp actually falls in the correct time window for the slot it is claiming.
+
+## The Block Import Trait
+
+When the import queue is instructed to import a block, it passes the block in question to a method provided by the [`BlockImport` trait](https://substrate.dev/rustdocs/v2.0.0-alpha.8/sp_consensus/block_import/trait.BlockImport.html). This `Block Import` trait provides the behavior of importing a block into the node's local state.
+
+One implementor of the `BlockImport` trait that is used in every Substrate node is the [`Client`](https://substrate.dev/rustdocs/v2.0.0-alpha.8/sc_service/client/index.html) which contains the node's entire block database. When a block is imported into the client it is added to the main database of blocks that the node knows about.
+
+## The Block Import Pipeline
+
+In the simplest cases, such as manual-seal nodes, blocks are imported directly into the client. But most consensus engines will need to perform additional verification on incoming blocks, update their own local state databases, or both. To allow consensus engines this opportunity, it is common to wrap the client in another struct which is also `BlockImport`. This nesting leads to the term "block import pipeline".
+
+An example of this wrapping is the [`PowBlockImport`](https://substrate.dev/rustdocs/v2.0.0-alpha.8/sc_consensus_pow/struct.PowBlockImport.html) which holds a reference to another type that is `BlockImport`. This allows the PoW consensus engine to do it's own import-related bookkeeping and then pass the block to the nested `BlockImport`, probably the client. This pattern is also demonstrated in the `AuraBlockImport`, `BabeBlockImport`, and `GrandpaBlockImport`.
```

Comment:
Worth adding links to Aura, Babe and Grandpa impls?

Reply:
I added the ones for Aura and Babe. For some reason the docs for the GrandpaBlocklmport https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/src/import.rs#L57 aren't published correctly as you can confirm by browsing around https://crates.parity.io/sc_finality_grandpa/index.html

Reply:
https://github.com/paritytech/substrate/pull/6068 makes the `GrandpaBlockImport` public, but this change was not included in alpha.8 so the question arises of how to link its docs. Here are some options.

1. Just don't link that one
2. Add the broken link knowing that it will be updated to a correct link when alpha.9 is released
3. Link to crates.parity.io

What do you think @joepetrowski

Reply:
@danforbes See here for discussion about linking to `GrandpaBlockImport`. I'll mark your comment below resolved to consolidate.

Reply:
Oh I think I may have even seen you chatting about this in Riot.

Reply:
Updated the links to crates.parity.io and added the `GrandpaBlockImport` link.